### PR TITLE
add "unpack_bits" function

### PR DESCRIPTION
**Pull request by arnej27959**
for discussion: is the name and syntax acceptable?

@havardpe 
@lesters 
@bratseth 
@jobergum

### CODE REVIEW

**Comment by bratseth** [comment]
Do we need the "_from_in8" suffix? Otherwise LGTM!

File: searchlib/src/test/java/com/yahoo/searchlib/rankingexpression/evaluation/EvaluationTestCase.java [hunk 1]
```diff
@@ -176,6 +176,28 @@ public void testBooleanEvaluation() {
...
+                               "tensor<int8>(a{},x[2]):{foo:[0,255],bar:[1,-8]}");
+
+        tester.assertEvaluates("tensor<int8>(a{},x[16]):{foo:[" +
+                               "0,0,0,0, 0,0,0,0," +
+                               "1,1,1,1, 1,1,1,1" +
+                               "],bar:[" +
+                               "1,0,0,0, 0,0,0,0," +
+                               "0,0,0,1, 1,1,1,1]}",
+                               "unpack_bits_from_int8(tensor0, int8, little)",
+                               "tensor<int8>(a{},x[2]):{foo:[0,255],bar:[1,-8]}");
```
Comment (jobergum) [inline searchlib/src/test/java/com/yahoo/searchlib/rankingexpression/evaluation/EvaluationTestCase.java:198]: I thought int8 was signed [-127,128] and not unsigned [0,255]?

Comment (jobergum) [reply searchlib/src/test/java/com/yahoo/searchlib/rankingexpression/evaluation/EvaluationTestCase.java:198]: Currently the "packing" is implemented in the embedder, maybe we can add the inversion as well (https://github.com/vespa-engine/vespa/blob/master/model-integration/src/test/java/ai/vespa/embedding/ColBertEmbedderTest.java)

Comment (arnej27959) [reply searchlib/src/test/java/com/yahoo/searchlib/rankingexpression/evaluation/EvaluationTestCase.java]: yes, 255 becomes -1, fixed in test.
not sure what you want with inversion, I can easily make this:
```
input: tensor<float>(d1[6],d2[8]):[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]]
packed: tensor<int8>(dt{},x[1]):{0:1.0, 1:5.0, 2:3.0, 3:127.0, 4:-128.0, 5:-1.0}
unpacked: tensor<float>(dt{},x[8]):{0:[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0], 1:[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0], 2:[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0], 3:[0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], 4:[1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 5:[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]}

input: tensor<float>(d1[2],d2[16]):[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]]
packed: tensor<int8>(dt{},x[2]):{0:[1.0, -128.0], 1:[5.0, 1.0]}
unpacked: tensor<float>(dt{},x[16]):{0:[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 1:[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]}
```

Comment (jobergum) [reply searchlib/src/test/java/com/yahoo/searchlib/rankingexpression/evaluation/EvaluationTestCase.java]: Thanks, yes, that is what I meant with inversion. That we have also `pack_into_int8` (uncertain about the name. The inspiration is from [numpy.packbits](https://numpy.org/doc/stable/reference/generated/numpy.packbits.html) and [numpy.unpackbits](https://numpy.org/doc/stable/reference/generated/numpy.unpackbits.html)

**Comment by lesters** [comment]
LGTM, great!

**Comment by jobergum** [comment]
Great! Thanks for adding the test showing compatibility with the packing done by the Colbert embedder. 

The numpy "equivalent" names are [unpackbits](https://numpy.org/doc/stable/reference/generated/numpy.unpackbits.html) and [packbits](https://numpy.org/doc/stable/reference/generated/numpy.packbits.html]. 

I think the name is fair since it doesn't make sense for other tensor cell types?

**Comment by arnej27959** [comment]
some changes and rebased now:
function name is just "unpack_bits"
type resolving checks that input has int8 cells (easy to work-around or lift restriction later)
default output cell type is "float", default endian is "big"

- unpack_bits(attribute(foo))
- unpack_bits(attribute(foo), float)
- unpack_bits(attribute(foo), float, big)

**Comment by jobergum** [comment]
Perfect @arnej27959 !

**Pull request merged**